appsrc/dcfile/pdftodc.man
.TH PDFTODC 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - Encapsulate PDF in DICOM"
.SH NAME
pdftodc \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Encapsulate PDF in DICOM
.SH SYNOPSIS
.HP 10
.B pdftodc
.so man1/gen.so
[
.B \-conceptname scheme code meaning version
]
[
.B \-v|verbose
]
.so man1/genin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B pdftodc
reads the named PDF input file and encapsulates in a new dicom file, creating an Encapsulated PDF Storage SOP
Class instance.
.LP
The output encoding of the file will be the default or explicitly
requested transfer syntax, as described in dcintro(1) under output
options.
.LP
During the translation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows the
specification of additional DICOM attributes, such as the patient's
name, necessary to make a complete and useful Encapsulated PDF instance.
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Display contents after creation, during replacement, and before writing.
.RE
.TP
.B \-conceptname scheme code meaning version
.RS
Fill in the ConceptNameCodeSequence attribute with the supplied values.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dctopdf(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS



CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/dcfile/pdftodc.cc
static const char *CopyrightIdentifier(void) { return "@(#)pdftodc.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#if USESTANDARDHEADERSWITHOUTEXTENSION == 1
#include <cctype>	// for isspace()
#else
#include <ctype.h>	// for isspace()
#endif

#if EMITUSINGSTDNAMESPACE == 1
using namespace std;
#endif

#include "attrtype.h"
#include "attrothr.h"
#include "attrmxls.h"
#include "attrseq.h"
#include "transynu.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "ioopt.h"
#include "dcopt.h"
#include "sopclu.h"

// addCodeSequenceItemToItemList() is copied from appsrc/dcfile/dcmulti.cc and probably should be factored out somewhere

static void addCodeSequenceItemToItemList(
	AttributeList *i,
	const char *csd,
	const char *cv,
	const char *cm,
	const char *csvn)
{
	Assert(i);
	//cm = lookUpAndFillInCodeMeaning(csd,cv,cm,csvn);
	if (csd  && strlen(csd))  (*i)+=new ShortStringAttribute(TagFromName(CodingSchemeDesignator),csd);
	if (cv   && strlen(cv))   (*i)+=new ShortStringAttribute(TagFromName(CodeValue),cv);
	if (cm   && strlen(cm))   (*i)+=new  LongStringAttribute(TagFromName(CodeMeaning),cm);
	if (csvn && strlen(csvn)) (*i)+=new ShortStringAttribute(TagFromName(CodingSchemeVersion),csvn);
}

// removeLeadingSpaces() is copied from libsrc/src/attrtyps.cc and probably should be factored out somewhere

static const char*
removeLeadingSpaces(const char *from)
{
	if (from) while (*from && isspace(*from)) ++from;
	return from;
}

int
main(int argc,char **argv)
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	InputOptions		input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	bool addConceptNameCodeSequence = false;
	const char *addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeDesignator;
	const char *addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeValue;
	const char *addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeMeaning;
	const char *addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeVersion;
	{
		const char *addConceptNameCodeSequenceArgs[4];
		int n = options.get("conceptname",addConceptNameCodeSequenceArgs,4);
//cerr << "addConceptNameCodeSequence returns " << dec << n << endl;
		if (n > 0) {
			if (n == 4) {
//cerr << "option: addConceptNameCodeSequence: addConceptNameCodeSequenceArgs[0] " << addConceptNameCodeSequenceArgs[0] << endl;
//cerr << "option: addConceptNameCodeSequence: addConceptNameCodeSequenceArgs[1] " << addConceptNameCodeSequenceArgs[1] << endl;
//cerr << "option: addConceptNameCodeSequence: addConceptNameCodeSequenceArgs[2] " << addConceptNameCodeSequenceArgs[2] << endl;
//cerr << "option: addConceptNameCodeSequence: addConceptNameCodeSequenceArgs[3] " << addConceptNameCodeSequenceArgs[3] << endl;
				addConceptNameCodeSequence=true;
				addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeDesignator = removeLeadingSpaces(addConceptNameCodeSequenceArgs[0]);
				             addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeValue = removeLeadingSpaces(addConceptNameCodeSequenceArgs[1]);
				           addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeMeaning = removeLeadingSpaces(addConceptNameCodeSequenceArgs[2]);
				   addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeVersion = removeLeadingSpaces(addConceptNameCodeSequenceArgs[3]);
			}
			else {
				cerr << "addreferenced needs 4 arguments, not " << dec << n << endl;
				bad=true;
			}
		}
	}

	input_options.done();
	dicom_output_options.done();
	options.done();

	InputOpenerFromOptions input_opener(
		options,input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< input_options.usage()
			<< dicom_output_options.usage()
			<< " [-conceptname scheme code meaning version]"
			<< " [-v|-verbose]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	istream in(input_opener);
	OurStreamPos pdfFileLength = OurStreamPos(0);
	in.seekg(0,ios::end);
	pdfFileLength = in.tellg();
	in.seekg(0,ios::beg);
	
	BinaryInputStream bin(in,NoEndian);

	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	TransferSyntax transfersyntax(
		dicom_output_options.transfersyntaxuid
		? dicom_output_options.transfersyntaxuid
		: DefaultTransferSyntaxUID);

	TextOutputStream log (cerr);
	if (verbose) log << "Converting ..." << endl;

	// various Type 1 and Type 2 attributes for mandatory Encapsulated Document modules ...

	list+=new PersonNameAttribute(TagFromName(PatientName),"^^^^");
	list+=new LongStringAttribute(TagFromName(PatientID));
	list+=new DateStringAttribute(TagFromName(PatientBirthDate));
	list+=new CodeStringAttribute(TagFromName(PatientSex));
	list+=new ShortStringAttribute(TagFromName(StudyID));
	list+=new DateStringAttribute(TagFromName(StudyDate));
	list+=new TimeStringAttribute(TagFromName(StudyTime));
	list+=new PersonNameAttribute(TagFromName(ReferringPhysicianName),"^^^^");
	list+=new ShortStringAttribute(TagFromName(AccessionNumber));
	list+=new IntegerStringAttribute(TagFromName(SeriesNumber),Int32(1));
	list+=new IntegerStringAttribute(TagFromName(InstanceNumber),Int32(1));
	list+=new DateStringAttribute(TagFromName(ContentDate));
	list+=new TimeStringAttribute(TagFromName(ContentTime));
	list+=new DateTimeStringAttribute(TagFromName(AcquisitionDateTime));

	list+=new UIStringAttribute(TagFromName(SOPClassUID),EncapsulatedPDFStorageSOPClassUID);
	list+=new CodeStringAttribute(TagFromName(Modality),"OT");
	list+=new CodeStringAttribute(TagFromName(ConversionType),"WSD");
	list+=new CodeStringAttribute(TagFromName(BurnedInAnnotation),"YES");

	list+=new LongStringAttribute(TagFromName(Manufacturer));
	
	list+=new ShortTextAttribute(TagFromName(DocumentTitle));
	
	SequenceAttribute *aConceptNameCodeSequence = new SequenceAttribute(TagFromName(ConceptNameCodeSequence));
	Assert(aConceptNameCodeSequence);
	list+=aConceptNameCodeSequence;
	if (addConceptNameCodeSequence) {
		AttributeList *iConceptNameCodeSequence = new AttributeList();
		Assert(iConceptNameCodeSequence);
		(*aConceptNameCodeSequence)+=iConceptNameCodeSequence;

		addCodeSequenceItemToItemList(iConceptNameCodeSequence,
			addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeDesignator,
			addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeValue,
			addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodeMeaning,
			addConceptNameCodeSequence_purposeOfReferenceCodeSequenceCodingSchemeVersion);
	}

	list+=new LongStringAttribute(TagFromName(MIMETypeOfEncapsulatedDocument),"application/pdf");
	list+=new OtherByteLargeNonPixelAttributeBase(TagFromName(EncapsulatedDocument),bin,OurStreamPos(0),pdfFileLength);		// pdfFileLength may be odd length; will be handled

	if (!usualManagedAttributeListWrite(list,dout,
		dicom_output_options,log,verbose)) return 1;

	return 0;
}

